tests/unit/memefficiency.tcl

```diff
@@ -397,6 +397,7 @@ run_solo {defrag} {
             r save ;# saving an rdb iterates over all the data / pointers
         } {OK}
 
+        if {[lindex [r config get io-threads] 1] == 1} {
```

Comment:
@oranagra io threads share arena with main thread is not good idea, performance drops by 5%.

> besides that, the `je_get_defrag_hint` makes the decision based on the average fragmentation in the allocation's arena, but then when it does `zmalloc_no_tcache` it uses a different one.

Does this also mean that we removed the fragment of the arena that was owned by the io thread and returned it to the main thread, it is still feasible for defragmentation, the fragment is still cleaned up.

> there are of course other cases where we keep a string (possibly an argv one) for a long time in the db (such as pubsub), but at least that one you resolved when you pinned pubsub to the main thread.

I just realized that the old io thread implementation still had the problems you mentioned, we just moved the fragment of Pubsub to main thread.

Reply:
>Does this also mean that we removed the fragment of the arena that was owned by the io thread and returned it to the main thread, it is still feasible for defragmentation, the fragment is still cleaned up.

yes. but it's a bit inefficient.

i'd like to focus on this:
>however, note that AFAIR in the majority of cases we do NOT use the data from argv directly in the db, we store a copy that we allocate in the main thread, see `sdsdup` in `dbAdd` and the code in `tryObjectEncoding` (specifically `trimStringObjectIfNeeded`).

so if that's true, my concerns about fragmentation and issues with the defragger are mostly dismissed, and we can maybe focus one case where the allocation from the io thread is retained.
let's try to map them, other than pubsub and modules, i suppose we also have an issue with BIGARG or some other string cases? can you confirm?

this means that this claim is actually inverted:
>Defragment: Since Pub/Sub clients still need to be handled by the main thread, memory allocation for Pub/Sub will occur on the main thread. In contrast, the memory for datasets created during normal operations is allocated by the IO threads. As a result, there is no fragmentation between Pub/Sub and datasets. Therefore, when IO threads are enabled, we skip the fragmentation test for Pub/Sub.

can you clear up what exactly failed in that test? i think it should be adjusted to pass rather than be skipped. and possibly add another test or two around the other cases in which an allocation from an IO thread is retained for long.

Reply:
> i'd like to focus on this:
> 
> > however, note that AFAIR in the majority of cases we do NOT use the data from argv directly in the db, we store a copy that we allocate in the main thread, see `sdsdup` in `dbAdd` and the code in `tryObjectEncoding` (specifically `trimStringObjectIfNeeded`).
> 
> so if that's true, my concerns about fragmentation and issues with the defragger are mostly dismissed, and we can maybe focus one case where the allocation from the io thread is retained. let's try to map them, other than pubsub and modules, i suppose we also have an issue with BIGARG or some other string cases? can you confirm?

1. string command: `append`(append to a nonexist string), `getset`, `mset` `set`.
   the value will be kept in db.
2. block related command.
    the keys will be kept in `c->db->blocking keys`.
3. watch command
    the keys kept in `c->db->watched_keys`.
4. [s]subscribe
    channel robj will be kept in `serverPubSubChannels`.

> 
> this means that this claim is actually inverted:
> 
> > Defragment: Since Pub/Sub clients still need to be handled by the main thread, memory allocation for Pub/Sub will occur on the main thread. In contrast, the memory for datasets created during normal operations is allocated by the IO threads. As a result, there is no fragmentation between Pub/Sub and datasets. Therefore, when IO threads are enabled, we skip the fragmentation test for Pub/Sub.
> 
> can you clear up what exactly failed in that test? i think it should be adjusted to pass rather than be skipped. and possibly add another test or two around the other cases in which an allocation from an IO thread is retained for l

because we use a few clients to subscribe a lot of channels to create a lot of fragments, but when the client execute the first subscribe command, the client will return to the main thread, so its subsequent memory allocation will be in the main thread, so under multi-threading, the pubsub command and the set command (used in the test) can't generate fragmentation.
we can use `setbit` to improve the test.

Reply:
> do you mean we do memory moves for these cases?

in some cases, tryObjectEncoding will fix it for us (move to the main thread arena), but maybe it's not the common case.

> because we use a few clients to subscribe a lot of channels to create a lot of fragments, but when the client execute the first subscribe command, the client will return to the main thread, so its subsequent memory allocation will be in the main thread, so under multi-threading, the pubsub command and the set command (used in the test) can't generate fragmentation.
we can use setbit to improve the test.

ok, so it's because we mix them. let's fix it maybe in the way you suggested instead of skipping the test.

but also maybe we need a dedicated test (or experiment) to see how multi-threading affects fragmentation (compared to single thread, before defrag is enabled), then see that defrag can fix it, and at what cost (maybe it's more than the 5% we lose when using a single arena).

we can leave that for after we resolve bigger problems...

Reply:
@oranagra i added a new test for io threads.
you're right, the fragmentation becomes more serious after io threads is turned on, the fragmentation rate of a single thread can be sorted to 5% in the test, but the fragmentation rate can only be sorted to 9% after 16 threads are turned on, the fragmentation of io threads becomes more dispersed and harder to be sorted.
we can be dealt with in the following PR.

Reply:
yes, but let's mention the concerns in the top comment.

Reply:
Updated the trade-off in top comment.

Reply:
i only see a comment about the defrag test. i think we need to describe the new fragmentation problems this brings (even if we already had them with the old IO threads).
and if there are some special cases concerns, edge cases, upper limits, or future plans.
i'm not sure which of the ones discussed here are worth mentioning, maybe none.

Reply:
sorry, I made a mistake and reverted it.
please have a look again.

Reply:
@oranagra the top comment updated.